File: client/Assets/LuaFramework/Lua/Common/Base64.lua

Base64 = {}
local this = Base64

function this.Init()
	local origin_str = "sY+N7PTnwM6bfAg4eaxkzCdhvJlUXZQSF82mDpK5WL/uEo31c0jiRtIHyqBOVGr9"
	this.code = {}
	for i = 1, #origin_str, 1 do
		table.insert(this.code, string.byte(string.sub(origin_str, i, i)))
	end

	this.reverse = {}
	for i = 1, 128, 1 do
		this.reverse[i] = 0
	end

	for i = 1, #this.code, 1 do
		this.reverse[this.code[i]] = i - 1
	end
end

this.Init()

function Base64.Encode( input )
	local input_len = #input
	local result_len = math.floor((input_len + 2) / 3) * 4
	local result = {}
	local index = 1
	local i = 3
	while i <= input_len do
		local tmp1 = base64bit:_lshift(input[i - 2], 16)
		local tmp2 = base64bit:_lshift(input[i - 1], 8)
		local val = base64bit:_or(base64bit:_or(tmp1, tmp2), input[i])
		result[index] = this.code[base64bit:_rshift(val, 18) + 1]
		index = index + 1
		result[index] = this.code[base64bit:_and(base64bit:_rshift(val, 12), 63) + 1]
		index = index + 1
		result[index] = this.code[base64bit:_and(base64bit:_rshift(val, 6), 63) + 1]
		index = index + 1
		result[index] = this.code[base64bit:_and(val, 63) + 1]
		index = index + 1
		i = i + 3
	end
	i = i - 2

	local mod = input_len % 3
	if mod == 1 then
		result[index] = this.code[base64bit:_rshift(input[i], 2) + 1]
		index = index + 1
		result[index] = this.code[base64bit:_and(base64bit:_lshift(input[i], 4), 63) + 1]
		index = index + 1
	elseif mod == 2 then
		result[index] = this.code[base64bit:_rshift(input[i], 2) + 1]
		index = index + 1
		result[index] = this.code[base64bit:_or(base64bit:_and(base64bit:_lshift(input[i], 4), 63), base64bit:_rshift(input[i+1], 4)) + 1]
		index = index + 1
		result[index] = this.code[base64bit:_and(base64bit:_lshift(input[i+1], 2), 63) + 1]
		index = index + 1
	end

	while index <= result_len do
		result[index] = string.byte("=")
		index = index + 1
	end
	return result
end

function Base64.Decode( input )
	local input_len = #input
	local len = math.floor(input_len / 4) * 3
	local tail = 0
	if len > 0 then
		tail = tail + (input[input_len] == string.byte("=") and 1 or 0)
		tail = tail + (input[input_len - 1] == string.byte("=") and 1 or 0)
		len = len - tail
	end
	local result = {}
	local i, index, search_len = 3, 1, input_len - tail
	while i < search_len do
		local tmp1 = base64bit:_lshift(this.reverse[input[i-2]], 18)
		local tmp2 = base64bit:_lshift(this.reverse[input[i-1]], 12)
		local tmp3 = base64bit:_lshift(this.reverse[input[i]], 6)
		local val = base64bit:_or(base64bit:_or(base64bit:_or(tmp1, tmp2), tmp3), this.reverse[input[i + 1]])
		result[index] = base64bit:_rshift(val, 16)
		index = index + 1
		result[index] = base64bit:_and(base64bit:_rshift(val, 8), 255)
		index = index + 1
		result[index] = base64bit:_and(val, 255)
		index = index + 1
		i = i + 4
	end
	i = i - 3
	if tail == 1 then
		local tmp1 = base64bit:_lshift(this.reverse[input[i + 1]], 12)
		local tmp2 = base64bit:_lshift(this.reverse[input[i+2]], 6)
		local val = base64bit:_or(base64bit:_or(tmp1, tmp2), this.reverse[input[i+3]])
		result[index] = base64bit:_rshift(val, 10)
		index = index + 1
		result[index] = base64bit:_and(base64bit:_rshift(val, 2), 255)
		index = index + 1
	elseif tail == 2 then
		val = base64bit:_or(base64bit:_lshift(this.reverse[input[i+1]], 6), this.reverse[input[i+2]])
		result[index] = base64bit:_rshift(val, 4)
		index = index + 1
	end
	return result
end

function Base64.EncodeString( str )
	if not str then return nil end
	local input = {}
	for i = 1, #str, 1 do
		table.insert(input, string.byte(string.sub(str, i, i)))
	end
	local output = Base64.Encode(input)
	local result = ''
	for k,v in ipairs(output) do
		result = result .. string.char(v)
	end
	return result
end

function Base64.DecodeString( str )
	if not str then return nil end
	local input = {}
    for i = 1, #str, 1 do
        table.insert(input, string.byte(string.sub(str, i, i)))
    end
    local output = Base64.Decode( input )
    local result = ''
    for k,v in ipairs(output) do
        result = result .. string.char(v)
    end
    return result
	-- body
end

-- print(this.DecodeString("6mJ+emeyfiZPemXRfiwqfkstazzRAkvya7aYARaNAmZPaNMNekMYaND="))

-- print(this.EncodeString("*6BB4837EB74329105EE4568DDA7DC67ED2CA2AD9"))
-- local inputs = {}
-- local str = "test"
-- for i = 1, 4, 1 do
-- 	table.insert(inputs, string.byte(string.sub(str, i, i)))
-- end

-- for k,v in ipairs(inputs) do
-- 	print(string.char(v))
-- end

-- print("=================================")

-- local enc = this.Encode(inputs)
-- for k,v in ipairs(enc) do
-- 	print(string.char(v))
-- end

-- print("=================================")
-- local dec = this.Decode(enc)
-- for k,v in ipairs(dec) do
-- 	print(string.char(v))
-- end

File: client/Assets/LuaFramework/Scripts/Common/Base64.cs
using System;
using System.Text;


// 2016 kingbird 
class Base64
{
    // 这个字符表跟标准的base64不同，只要和服务器约定一样的序列就可以使用，非标准序列可以做到更安全的加密
    byte[] code = Encoding.ASCII.GetBytes("sY+N7PTnwM6bfAg4eaxkzCdhvJlUXZQSF82mDpK5WL/uEo31c0jiRtIHyqBOVGr9");
    byte[] reverse = new byte[128];

    static Base64 inst = new Base64();

    private Base64()
    {
        for(int i = 0; i < code.Length; i++)
        {
            reverse[code[i]] = (byte)i;
        }
    }

    public static Base64 Inst
    {
        get { return inst; }
    }

    // 假定byte长度是8位了
    public byte[] Encode(byte[] input)
    {
        byte[] ret = new byte[((input.Length + 2)/3) * 4];
        int index = 0;
        int i;
        for (i = 2 ; i < input.Length; i += 3)
        {
            int val = (input[i - 2] << 16) | (input[i - 1] << 8) | input[i];
            ret[index++] = code[val >> 18];
            ret[index++] = code[(val >> 12) & 0x3f];
            ret[index++] = code[(val >> 6) & 0x3F];
            ret[index++] = code[val & 0x3F];
        }
        i -= 2;
        switch(input.Length % 3)
        {
            case 1:
                ret[index++] = code[input[i] >> 2];
                ret[index++] = code[(input[i] << 4) & 0x3F];
                break;
            case 2:
                ret[index++] = code[input[i] >> 2];
                ret[index++] = code[(input[i] << 4) & 0x3F | (input[i+1] >> 4)];
                ret[index++] = code[(input[i+1] << 2) & 0x3F];
                break;
            default:break;
        }

        while (index < ret.Length)
        {
            ret[index++] = (byte)'=';
        }
        return ret;
    } 

    public byte[] Decode(byte[] input)
    {
        int len = (input.Length / 4) * 3;
        int tail = 0;
        if (len > 0)
        {
            tail += input[input.Length - 1] == (byte)'=' ? 1 : 0;
            tail += input[input.Length - 2] == (byte)'=' ? 1 : 0;
            len -= tail;
        }
        byte[] ret = new byte[len];

        int i, index = 0, serach_len = input.Length - tail;
        for (i = 3; i < serach_len; i += 4)
        {
            int val = (reverse[input[i - 3]] << 18) | (reverse[input[i - 2]] << 12) | (reverse[input[i - 1]] << 6) | reverse[input[i]];
            ret[index++] = (byte)(val >> 16);
            ret[index++] = (byte)((val >> 8) & 0xFF);
            ret[index++] = (byte)(val &0xFF);
        }
        i -= 3;
        switch (tail)
        {
            case 1:
                int val = (reverse[input[i]] << 12) | (reverse[input[i+1]] << 6) | reverse[input[i+2]];
                ret[index++] = (byte)(val >> 10);
                ret[index++] = (byte)((val >> 2) & 0xFF);
                break;
            case 2:
                val = (reverse[input[i]] << 6) | reverse[input[i + 1]];
                ret[index++] = (byte)(val >> 4);
                break;
            default: break;
        }

        return ret;
    }
}

